# Interactive Reference for pandas

## Part 0: Background

### What is pandas?


"pandas is an open source, BSD-licensed library providing high-performance, easy-to-use data structures and data analysis tools for the Python programming language."

Pandas provides Dataframes, an easy-to-use and efficient tool for interacting with medium to large amounts of tabular data.

This workbook will attempt to give an overview of the funtionality of pandas as well as provide examples of common operations which you can refer back to in your data science adventures.

## Part 1: Loading and Examining Data

For the duration of this workbook we're going to use a simple .csv file found online which provides 58 examples of emails marked as Spam or Not Spam. This kind of data can be used for training classifiers for spam detection.

We'll start by downloading the data using a simple wget command. Remeber that anything following a '!' in a jupyter notebook code block will be executed as a shell command.

### A guideline for file sizes:

The original creator of pandas recommends always having at least 5-10x the size of your data in available memory, to deal with the overhead of using pandas and ensure that you won't experience OOM errors. [Read his explanation and more about the development of pandas here.](http://wesmckinney.com/blog/apache-arrow-pandas-internals/)

In [128]:
! wget https://www.openml.org/data/get_csv/44/dataset_44_spambase.csv

--2018-09-23 09:53:44--  https://www.openml.org/data/get_csv/44/dataset_44_spambase.csv
Resolving www.openml.org (www.openml.org)... 131.155.11.58
Connecting to www.openml.org (www.openml.org)|131.155.11.58|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/plain]
Saving to: ‘dataset_44_spambase.csv.1’

dataset_44_spambase     [        <=>         ] 682.99K   354KB/s    in 1.9s    

2018-09-23 09:53:47 (354 KB/s) - ‘dataset_44_spambase.csv.1’ saved [699384]



### Great, now we have our data!

Next, we'll start by importing Pandas, reading the .csv and then having a look at our data.

In [129]:
# Import pandas, we'll use it a lot so give it the shorter name 'pd'.
import pandas as pd

In [130]:
# Read the .csv file into a pandas dataframe called df.
df = pd.read_csv('dataset_44_spambase.csv')

In [131]:
# Check what kind of file was returned by pd.read_csv()
print("Object type: ", type(df))

# Check out the dimensions of the 
print("Dimensions of the data: ", df.shape)

Object type:  <class 'pandas.core.frame.DataFrame'>
Dimensions of the data:  (4601, 58)


So now we know that we have a Pandas Dataframe with 4601 rows and 58 columns. We can explore further to see what these columns actually mean by examing the header of the dataframe.

In [132]:
# Casting the dataframe to a list will give us a list of all the column names:
column_names = list(df)

# Print the column names one at a time
for column_name in column_names: print(column_name)

word_freq_make
word_freq_address
word_freq_all
word_freq_3d
word_freq_our
word_freq_over
word_freq_remove
word_freq_internet
word_freq_order
word_freq_mail
word_freq_receive
word_freq_will
word_freq_people
word_freq_report
word_freq_addresses
word_freq_free
word_freq_business
word_freq_email
word_freq_you
word_freq_credit
word_freq_your
word_freq_font
word_freq_000
word_freq_money
word_freq_hp
word_freq_hpl
word_freq_george
word_freq_650
word_freq_lab
word_freq_labs
word_freq_telnet
word_freq_857
word_freq_data
word_freq_415
word_freq_85
word_freq_technology
word_freq_1999
word_freq_parts
word_freq_pm
word_freq_direct
word_freq_cs
word_freq_meeting
word_freq_original
word_freq_project
word_freq_re
word_freq_edu
word_freq_table
word_freq_conference
char_freq_%3B
char_freq_%28
char_freq_%5B
char_freq_%21
char_freq_%24
char_freq_%23
capital_run_length_average
capital_run_length_longest
capital_run_length_total
class


In [133]:
# We can also check the number of unique items per row with the command df[col_name].nunique
for column_name in column_names:
    print(column_name, " : ", df[column_name].nunique())

word_freq_make  :  142
word_freq_address  :  171
word_freq_all  :  214
word_freq_3d  :  43
word_freq_our  :  255
word_freq_over  :  141
word_freq_remove  :  173
word_freq_internet  :  170
word_freq_order  :  144
word_freq_mail  :  245
word_freq_receive  :  113
word_freq_will  :  316
word_freq_people  :  158
word_freq_report  :  133
word_freq_addresses  :  118
word_freq_free  :  253
word_freq_business  :  197
word_freq_email  :  229
word_freq_you  :  575
word_freq_credit  :  148
word_freq_your  :  401
word_freq_font  :  99
word_freq_000  :  164
word_freq_money  :  143
word_freq_hp  :  395
word_freq_hpl  :  281
word_freq_george  :  240
word_freq_650  :  200
word_freq_lab  :  156
word_freq_labs  :  179
word_freq_telnet  :  128
word_freq_857  :  106
word_freq_data  :  184
word_freq_415  :  110
word_freq_85  :  177
word_freq_technology  :  159
word_freq_1999  :  188
word_freq_parts  :  53
word_freq_pm  :  163
word_freq_direct  :  125
word_freq_cs  :  108
word_freq_meeting  :  186
word_freq_

### The _df.describe()_ function:

This returns a summary of all the numerical data in your dataframe on a column-by-column basis. It can be great for getting a quick idea of what your data looks like.


In [134]:
df.describe()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
count     4601.000000        4601.000000    4601.000000   4601.000000   
mean         0.104553           0.213015       0.280656      0.065425   
std          0.305358           1.290575       0.504143      1.395151   
min          0.000000           0.000000       0.000000      0.000000   
25%          0.000000           0.000000       0.000000      0.000000   
50%          0.000000           0.000000       0.000000      0.000000   
75%          0.000000           0.000000       0.420000      0.000000   
max          4.540000          14.280000       5.100000     42.810000   

       word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
count    4601.000000     4601.000000       4601.000000         4601.000000   
mean        0.312223        0.095901          0.114208            0.105295   
std         0.672513        0.273824          0.391441            0.401071   
min         0.000000        0.000000          0.000000            0.000000   
25%         0.000000        0.000000          0.000000            0.000000   
50%         0.000000        0.000000          0.000000            0.000000   
75%         0.380000        0.000000          0.000000            0.000000   
max        10.000000        5.880000          7.270000           11.110000   

       word_freq_order  word_freq_mail     ...       char_freq_%3B  \
count      4601.000000     4601.000000     ...         4601.000000   
mean          0.090067        0.239413     ...            0.038575   
std           0.278616        0.644755     ...            0.243471   
min           0.000000        0.000000     ...            0.000000   
25%           0.000000        0.000000     ...            0.000000   
50%           0.000000        0.000000     ...            0.000000   
75%           0.000000        0.160000     ...            0.000000   
max           5.260000       18.180000     ...            4.385000   

       char_freq_%28  char_freq_%5B  char_freq_%21  char_freq_%24  \
count    4601.000000    4601.000000    4601.000000    4601.000000   
mean        0.139030       0.016976       0.269071       0.075811   
std         0.270355       0.109394       0.815672       0.245882   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.065000       0.000000       0.000000       0.000000   
75%         0.188000       0.000000       0.315000       0.052000   
max         9.752000       4.081000      32.478000       6.003000   

       char_freq_%23  capital_run_length_average  capital_run_length_longest  \
count    4601.000000                 4601.000000                 4601.000000   
mean        0.044238                    5.191515                   52.172789   
std         0.429342                   31.729449                  194.891310   
min         0.000000                    1.000000                    1.000000   
25%         0.000000                    1.588000                    6.000000   
50%         0.000000                    2.276000                   15.000000   
75%         0.000000                    3.706000                   43.000000   
max        19.829000                 1102.500000                 9989.000000   

       capital_run_length_total        class  
count               4601.000000  4601.000000  
mean                 283.289285     0.394045  
std                  606.347851     0.488698  
min                    1.000000     0.000000  
25%                   35.000000     0.000000  
50%                   95.000000     0.000000  
75%                  266.000000     1.000000  
max                15841.000000     1.000000  

[8 rows x 58 columns]

Note that any of the statistics seen above can be calculated on their own on a column-by-column basis, either across the whole datafram or across a specific range / selection of columns.

In [135]:
# Calculate the sum of all columns and display the first 10
df.sum()[:10]

word_freq_make         481.05
word_freq_address      980.08
word_freq_all         1291.30
word_freq_3d           301.02
word_freq_our         1436.54
word_freq_over         441.24
word_freq_remove       525.47
word_freq_internet     484.46
word_freq_order        414.40
word_freq_mail        1101.54
dtype: float64

In [136]:
# Calculate the average of the capital_rnu_length_longest column.
df['capital_run_length_longest'].mean()

52.17278852423386

## Part 2: Retrieving Data

It's rare that you'll want the data set in its entirety, so we'll use some of pandas' functionality to only return a subset of the rows and/or columns in our dataframe.

Columns can be indexed by their label.

In [137]:
df['word_freq_remove']

0       0.00
1       0.21
2       0.19
3       0.31
4       0.31
5       0.00
6       0.00
7       0.00
8       0.30
9       0.38
10      0.96
11      0.25
12      0.00
13      0.90
14      0.00
15      0.42
16      0.00
17      0.00
18      0.18
19      0.00
20      0.00
21      0.15
22      0.00
23      0.00
24      0.00
25      0.15
26      0.00
27      1.66
28      0.00
29      0.65
        ... 
4571    0.00
4572    0.00
4573    0.00
4574    0.00
4575    0.00
4576    0.00
4577    0.00
4578    0.00
4579    0.00
4580    0.00
4581    0.00
4582    0.00
4583    0.00
4584    0.00
4585    0.00
4586    0.00
4587    0.00
4588    0.00
4589    0.00
4590    0.00
4591    0.00
4592    0.00
4593    0.00
4594    0.00
4595    0.00
4596    0.00
4597    0.00
4598    0.00
4599    0.00
4600    0.00
Name: word_freq_remove, Length: 4601, dtype: float64

Multiple columns can be retrieved by providing a list of column names (Remember to use a double set of square brackets):

In [138]:

df[['word_freq_over','word_freq_remove','word_freq_internet']]

word_freq_over  word_freq_remove  word_freq_internet
0               0.00              0.00                0.00
1               0.28              0.21                0.07
2               0.19              0.19                0.12
3               0.00              0.31                0.63
4               0.00              0.31                0.63
5               0.00              0.00                1.85
6               0.00              0.00                0.00
7               0.00              0.00                1.88
8               0.00              0.30                0.00
9               0.32              0.38                0.00
10              0.00              0.96                0.00
11              0.25              0.25                0.00
12              0.00              0.00                0.00
13              0.00              0.90                0.00
14              0.35              0.00                0.35
15              0.00              0.42                0.00
16              0.00              0.00                0.00
17              0.00              0.00                0.00
18              0.00              0.18                0.00
19              0.31              0.00                0.00
20              0.00              0.00                0.00
21              0.05              0.15                0.02
22              0.00              0.00                0.00
23              0.00              0.00                0.00
24              0.00              0.00                0.00
25              0.05              0.15                0.02
26              0.00              0.00                0.00
27              0.00              1.66                0.00
28              0.00              0.00                0.00
29              0.00              0.65                0.00
...              ...               ...                 ...
4571            0.23              0.00                0.00
4572            0.00              0.00                0.00
4573            0.18              0.00                0.00
4574            0.00              0.00                0.00
4575            0.00              0.00                0.00
4576            0.00              0.00                0.00
4577            0.00              0.00                0.00
4578            0.00              0.00                0.00
4579            0.00              0.00                0.00
4580            0.00              0.00                0.00
4581            0.51              0.00                0.00
4582            0.00              0.00                0.00
4583            0.00              0.00                0.00
4584            0.22              0.00                0.00
4585            0.00              0.00                0.00
4586            0.00              0.00                0.00
4587            0.00              0.00                0.00
4588            0.00              0.00                0.00
4589            0.00              0.00                0.00
4590            0.00              0.00                0.00
4591            0.00              0.00                0.00
4592            0.00              0.00                0.00
4593            0.00              0.00                0.00
4594            0.00              0.00                0.00
4595            0.00              0.00                0.00
4596            0.31              0.00                0.00
4597            0.00              0.00                0.00
4598            0.00              0.00                0.00
4599            0.00              0.00                0.00
4600            0.00              0.00                0.00

[4601 rows x 3 columns]

Entire rows can be retrieved range or by criteria. Use df[:][low:high] to select all columns for a given row.

In [139]:
df[:][14:16]

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
14             0.0               0.00           1.42           0.0   
15             0.0               0.42           0.42           0.0   

    word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
14           0.71            0.35              0.00                0.35   
15           1.27            0.00              0.42                0.00   

    word_freq_order  word_freq_mail  ...    char_freq_%3B  char_freq_%28  \
14              0.0            0.71  ...              0.0          0.102   
15              0.0            1.27  ...              0.0          0.063   

    char_freq_%5B  char_freq_%21  char_freq_%24  char_freq_%23  \
14            0.0          0.357          0.000            0.0   
15            0.0          0.572          0.063            0.0   

    capital_run_length_average  capital_run_length_longest  \
14                       1.971                          24   
15                       5.659                          55   

    capital_run_length_total  class  
14                       205      1  
15                       249      1  

[2 rows x 58 columns]

Rows can also be selected based on a query. 

In [140]:
df.query('word_freq_make > 0.10')

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1               0.21               0.28           0.50           0.0   
8               0.15               0.00           0.46           0.0   
30              1.17               0.00           0.00           0.0   
40              0.30               0.00           0.00           0.0   
45              0.15               0.45           1.05           0.0   
46              0.18               0.00           0.18           0.0   
47              0.49               0.00           0.99           0.0   
48              0.46               0.30           0.46           0.0   
49              0.46               0.46           0.26           0.0   
51              0.73               0.36           1.09           0.0   
54              0.73               0.36           0.73           0.0   
61              0.23               0.00           0.47           0.0   
64              0.12               0.24           0.12           0.0   
65              0.66               0.00           0.66           0.0   
68              0.30               0.00           0.61           0.0   
77              0.75               0.18           0.37           0.0   
91              0.47               0.31           0.47           0.0   
92              0.59               0.44           0.29           0.0   
93              0.59               0.44           0.29           0.0   
94              0.46               0.00           0.46           0.0   
99              1.24               0.41           1.24           0.0   
114             0.12               1.76           0.63           0.0   
122             0.50               0.46           0.34           0.0   
125             0.16               0.00           0.67           0.0   
126             0.50               0.00           0.50           0.0   
129             0.41               0.00           0.41           0.0   
137             1.00               0.00           0.33           0.0   
140             1.00               0.00           0.33           0.0   
142             0.50               0.46           0.34           0.0   
146             1.00               0.00           0.33           0.0   
...              ...                ...            ...           ...   
4418            0.59               0.59           0.00           0.0   
4424            0.40               0.00           0.81           0.0   
4450            0.30               0.00           0.30           0.0   
4451            0.12               0.00           0.12           0.0   
4458            0.25               0.00           0.51           0.0   
4460            0.16               0.16           0.16           0.0   
4461            0.95               0.00           0.00           0.0   
4471            1.42               0.00           0.00           0.0   
4472            0.76               0.00           0.76           0.0   
4473            0.26               0.00           0.00           0.0   
4481            1.88               0.00           0.00           0.0   
4487            0.39               0.00           0.98           0.0   
4489            0.35               0.00           0.71           0.0   
4500            0.78               0.00           0.00           0.0   
4507            0.23               0.00           0.00           0.0   
4514            0.88               0.00           0.00           0.0   
4523            0.42               0.00           0.42           0.0   
4554            0.33               0.00           0.00           0.0   
4555            0.17               0.00           0.68           0.0   
4557            0.69               0.00           0.00           0.0   
4558            0.16               0.00           0.32           0.0   
4563            0.54               0.00           0.54           0.0   
4567            0.14               0.00           0.14           0.0   
4568            0.67               0.00       

These queries can be chained for further refinement:

In [141]:
df.query('word_freq_make > 0.10').query('word_freq_address > 0.60').query('word_freq_our < 0.03')

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
234             0.84               0.84           0.00           0.0   
372             0.20               0.81           0.61           0.0   
516             0.26               0.72           0.85           0.0   
549             0.33               0.67           0.00           0.0   
1153            0.74               0.74           0.74           0.0   
1397            0.62               0.62           0.00           0.0   
1531            0.63               0.63           0.63           0.0   
1630            4.54               4.54           0.00           0.0   
3247            0.34               1.36           0.00           0.0   
3470            1.17               3.52           0.00           0.0   
4118            0.66               0.66           0.66           0.0   
4132            0.63               0.63           0.00           0.0   
4257            1.47               1.47           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
234             0.0            0.00              1.69                0.00   
372             0.0            0.00              0.00                0.00   
516             0.0            0.19              0.06                0.33   
549             0.0            0.00              0.67                0.33   
1153            0.0            0.00              0.37                0.00   
1397            0.0            1.86              0.00                0.00   
1531            0.0            0.00              0.63                0.63   
1630            0.0            0.00              0.00                0.00   
3247            0.0            0.00              0.00                0.00   
3470            0.0            0.00              0.00                0.00   
4118            0.0            0.00              0.00                0.00   
4132            0.0            0.00              0.00                0.00   
4257            0.0            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...    char_freq_%3B  char_freq_%28  \
234              0.84            0.84  ...              0.0          0.000   
372              0.20            0.00  ...              0.0          0.000   
516              0.72            0.46  ...              0.0          0.131   
549              0.33            0.33  ...              0.0          0.000   
1153             0.37            1.12  ...              0.0          0.000   
1397             0.00            0.00  ...              0.0          0.327   
1531             0.63            0.00  ...              0.0          0.000   
1630             0.00            0.00  ...              0.0          0.000   
3247             0.34            1.70  ...              0.0          0.244   
3470             0.00            1.17  ...              0.0          0.000   
4118             0.00            1.33  ...              0.0          0.000   
4132             0.00            0.63  ...              0.0          0.204   
4257             0.00            1.47  ...              0.0          0.000   

      char_freq_%5B  char_freq_%21  char_freq_%24  char_freq_%23  \
234           0.000          1.113          0.278          0.092   
372           0.000          0.831          0.338          0.030   
516           0.000          0.101          0.101          0.202   
549           0.000          0.000          0.048          0.000   
1153          0.061          0.061          0.122          0.000   
1397          0.000          2.295          0.218          0.000   
1531          0.000          0.398          0.000          0.000   
1630          0.000          3.076          0.000          0.000   
3247          0.000          0.000          0.000          0.000   
3470          0.000          0.414          0.000          0.000   
4118          0.000          0.117          0.000          0.000   
4132     

### Sampling from Dataframes

Sometimes the dataset is too big, or you want to prototype your model with a smaller amount of data to save time. Pandas provides a sampling method to handle these instances. the number of samples required can be provided as an integer or a fraction relative to the total number of samples. 

Including the term 'replace=True' will cause the samples to be drawn with replacement.

In [142]:
df.sample(n=5)

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
4218            0.00               0.00           0.71           0.0   
2130            0.26               0.00           0.26           0.0   
374             0.00               0.00           0.00           0.0   
1315            0.27               0.27           0.55           0.0   
3008            0.00               0.00           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
4218           0.00            0.00              0.00                0.00   
2130           0.00            0.00              0.00                0.00   
374            1.48            0.74              1.48                0.00   
1315           0.27            0.27              0.00                1.39   
3008           0.13            0.26              0.00                0.00   

      word_freq_order  word_freq_mail  ...    char_freq_%3B  char_freq_%28  \
4218             0.00            0.00  ...            0.000          0.000   
2130             0.00            0.00  ...            0.123          0.041   
374              0.00            0.74  ...            0.000          0.000   
1315             0.27            0.83  ...            0.000          0.279   
3008             0.00            0.13  ...            0.135          0.101   

      char_freq_%5B  char_freq_%21  char_freq_%24  char_freq_%23  \
4218            0.0          0.342          0.000            0.0   
2130            0.0          0.041          0.000            0.0   
374             0.0          0.108          0.000            0.0   
1315            0.0          2.001          0.093            0.0   
3008            0.0          0.000          0.000            0.0   

      capital_run_length_average  capital_run_length_longest  \
4218                       1.000                           1   
2130                       1.517                           4   
374                        2.346                          12   
1315                       3.706                          63   
3008                       1.915                          19   

      capital_run_length_total  class  
4218                        31      0  
2130                        44      0  
374                         61      1  
1315                       341      1  
3008                       387      0  

[5 rows x 58 columns]

In [143]:
df.sample(frac=0.0008, replace=True)

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
2887            0.00               0.00           0.00           0.0   
1058            0.10               0.10           0.71           0.0   
812             0.08               0.08           0.76           0.0   
2559            0.00               0.00           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
2887           0.00            0.00              0.00                0.00   
1058           0.60            0.30              0.40                0.10   
812            0.85            1.02              0.25                0.17   
2559           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...    char_freq_%3B  char_freq_%28  \
2887             0.00            0.00  ...              0.0          0.000   
1058             1.42            0.81  ...              0.0          0.000   
812              0.59            0.08  ...              0.0          0.065   
2559             0.00            0.00  ...              0.0          0.000   

      char_freq_%5B  char_freq_%21  char_freq_%24  char_freq_%23  \
2887          0.000          0.000          0.000          0.000   
1058          0.264          0.977          0.397          0.033   
812           0.000          0.408          0.118          0.013   
2559          0.000          0.000          0.000          0.000   

      capital_run_length_average  capital_run_length_longest  \
2887                       1.000                           1   
1058                       3.166                          56   
812                        7.550                         669   
2559                       1.533                           5   

      capital_run_length_total  class  
2887                         6      0  
1058                      1045      1  
812                       1412      1  
2559                        23      0  

[4 rows x 58 columns]

### Integer-based indexing using iloc

Pandas also provides the iloc function to retrieve rowsdata purely based on integer indices.

In [144]:
df.iloc[5:8]

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
5             0.0                0.0            0.0           0.0   
6             0.0                0.0            0.0           0.0   
7             0.0                0.0            0.0           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
5           1.85             0.0               0.0                1.85   
6           1.92             0.0               0.0                0.00   
7           1.88             0.0               0.0                1.88   

   word_freq_order  word_freq_mail  ...    char_freq_%3B  char_freq_%28  \
5              0.0            0.00  ...              0.0          0.223   
6              0.0            0.64  ...              0.0          0.054   
7              0.0            0.00  ...              0.0          0.206   

   char_freq_%5B  char_freq_%21  char_freq_%24  char_freq_%23  \
5            0.0          0.000          0.000            0.0   
6            0.0          0.164          0.054            0.0   
7            0.0          0.000          0.000            0.0   

   capital_run_length_average  capital_run_length_longest  \
5                       3.000                          15   
6                       1.671                           4   
7                       2.450                          11   

   capital_run_length_total  class  
5                        54      1  
6                       112      1  
7                        49      1  

[3 rows x 58 columns]

### Sorting data

We can use the sort function to order our dataframe based on a specific numeric column. The default is low to high, so use ascending=False to get the descending ordering. 

In [156]:
# Get the 10 rows with the highest value of word_freq_make
df.sort_values('word_freq_make', ascending=False)[:10]

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1630            4.54               4.54            0.0           0.0   
4378            4.34               0.00            0.0           0.0   
4292            4.00               0.00            1.0           0.0   
3186            3.94               0.00            0.0           0.0   
4065            3.84               0.00            0.0           0.0   
3479            3.03               0.00            0.0           0.0   
3932            2.85               0.00            0.0           0.0   
2215            2.77               0.00            0.0           0.0   
2077            2.77               0.00            0.0           0.0   
566             2.43               0.00            0.0           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1630            0.0             0.0               0.0                 0.0   
4378            0.0             0.0               0.0                 0.0   
4292            1.0             0.0               0.0                 0.0   
3186            0.0             0.0               0.0                 0.0   
4065            0.0             0.0               0.0                 0.0   
3479            0.0             0.0               0.0                 0.0   
3932            0.0             0.0               0.0                 0.0   
2215            0.0             0.0               0.0                 0.0   
2077            0.0             0.0               0.0                 0.0   
566             0.0             0.0               0.0                 0.0   

      word_freq_order  word_freq_mail     ...       char_freq_%28  \
1630              0.0             0.0     ...               0.000   
4378              0.0             0.0     ...               0.000   
4292              0.0             0.0     ...               0.168   
3186              0.0             0.0     ...               0.000   
4065              0.0             0.0     ...               0.645   
3479              0.0             0.0     ...               0.609   
3932              0.0             0.0     ...               0.465   
2215              0.0             0.0     ...               0.000   
2077              0.0             0.0     ...               0.000   
566               0.0             0.0     ...               0.000   

      char_freq_%5B  char_freq_%21  char_freq_%24  char_freq_%23  \
1630            0.0          3.076            0.0            0.0   
4378            0.0          0.000            0.0            0.0   
4292            0.0          0.168            0.0            0.0   
3186            0.0          0.000            0.0            0.0   
4065            0.0          0.000            0.0            0.0   
3479            0.0          0.000            0.0            0.0   
3932            0.0          0.000            0.0            0.0   
2215            0.0          0.000            0.0            0.0   
2077            0.0          0.213            0.0            0.0   
566             0.0          0.000            0.0            0.0   

      capital_run_length_average  capital_run_length_longest  \
1630                       1.428                           4   
4378                       1.750                           4   
4292                       1.459                          10   
3186                       1.800                           5   
4065                       1.000                           1   
3479                       1.181                           3   
3932                       1.250                           3   
2215                       1.100                           2   
2077                       1.750                           6   
566                        3.666                          13   

      capital_run_length_total  class  new_feature  
1630                        10      1          NaN  
4378                        14      0        

## Part 3: Adding Data

Sometimes you'll want to add a new column calculated as a function of other columns (feature engineering).

In [145]:
df['new_feature'] = df.word_freq_all / df.word_freq_over

df['new_feature'][:10]

0         inf
1    1.785714
2    3.736842
3         NaN
4         NaN
5         NaN
6         NaN
7         NaN
8         inf
9    2.406250
Name: new_feature, dtype: float64

## Part 4: Cleaning Data

One of the most immportant steps in data cleaning is to handle missing data. Pandas provides functions for either replacing missing values or dropping rows with missing data from the data.

In [146]:
# Replace NaN values with 0.00
df['new_feature'].fillna(0.00)[1:10]

1    1.785714
2    3.736842
3    0.000000
4    0.000000
5    0.000000
6    0.000000
7    0.000000
8         inf
9    2.406250
Name: new_feature, dtype: float64

We can also replace specific values (often used to handle inf)

In [147]:
# Replace NaN with 0.00, replace inf with 100.0
df['new_feature'].fillna(0.00).replace(float('inf'), 100.0)[1:10]

1      1.785714
2      3.736842
3      0.000000
4      0.000000
5      0.000000
6      0.000000
7      0.000000
8    100.000000
9      2.406250
Name: new_feature, dtype: float64

In [148]:
# Print the shape of the table 
print("Dataframe shape before cleaning: ", df.shape)

# Drop all of the rows containing NaN values:
clean_df = df.dropna()

# print the Size again
print("Dataframe shape after cleaning: ", clean_df.shape)


Dataframe shape before cleaning:  (4601, 59)
Dataframe shape after cleaning:  (2118, 59)


## Part 5: Other Useful Snippets

We can use the memory_usage() funtion to view the memory usage of each column, returned in bytes.

In [153]:
# view the memory usage of the first 10 columns (in bytes)
df.memory_usage()[:10]

Index                    80
word_freq_make        36808
word_freq_address     36808
word_freq_all         36808
word_freq_3d          36808
word_freq_our         36808
word_freq_over        36808
word_freq_remove      36808
word_freq_internet    36808
word_freq_order       36808
dtype: int64